src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -315,12 +321,36 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
             try!(fingerprint::append_current_dir(&dep_info_loc, &cwd));
         }
 
+        // Hard link the expected executable from the metadata hashed one
+        for &(ref filename, ref bin_name, _linkable) in filenames.iter() {
+            if filename != bin_name {
+                let src = root.join(filename);
+                let dst = root.join(bin_name);
+                // This may have been a `cargo rustc` command which changes the
+                // output, so the source may not actually exist.
+                if !src.exists() {
+                    continue
+                }
+                if dst.exists() {
+                    try!(fs::remove_file(&dst).chain_error(|| {
+                        human(format!("failed to remove: {}", dst.display()))
+                    }));
+                }
+                try!(fs::hard_link(&src, &dst)
+                     .or_else(|_| fs::copy(&src, &dst).map(|_| ()))
+                     .chain_error(|| {
+                         human(format!("failed to link or copy `{}` to `{}`",
+                                       src.display(), dst.display()))
+                }));
```

Comment:
Very interesting. This code is non-deterministically failing due to parallelism + the non-atomicity of the remove_file, hard_link, and  fs_copy on the test which is compiling two versions of a crate (use_a_spec_to_select).
Here is the ordering that causes a failure.

thread 1) foo-abcdef
thread 2) foo-123456

1) Checks if foo exists [no]
2) Checks if foo exists [no]
1) hard links foo(inode x) -> foo-abcdef(inode x)
2) Attempts to hard link foo -> foo-123456 (fails due to foo existing)
2) Attempts to copy foo-123456 -> foo (succeeds, overwriting inode x and both versions of foo).

Now foo, foo-123456, and foo-abcdef all have the same contents.
We need to somehow prevent either
1) Multiple versions of the same crate being compiled concurrently
2) Multiple versions attempting hard link concurrently.